ldap/servers/plugins/alias_entries/alias-entries.c

```diff
@@ -0,0 +1,123 @@
+ *
+ * License: GPL (version 3 or any later version).
+ * See LICENSE for details.
+ * END COPYRIGHT BLOCK **/
+
+#include "alias-entries.h"
+
+static Slapi_ComponentId *plugin_id = NULL;
+Slapi_PluginDesc srchpdesc = { PLUGINNAME, PLUGINVNDR, PLUGINVERS, PLUGINDESC };
+
+static Slapi_DN *
+alias_get_next_dn(Slapi_DN *sdn)
+{
+    Slapi_PBlock *pb = NULL;
+    Slapi_Entry **e;
+    Slapi_DN *entry_sdn = NULL;
+    int derefNever = LDAP_DEREF_NEVER;
+    char *attrs[] = { "aliasedObjectName", NULL };
+
+    /* search dn to check if it is an alias */
+    pb = slapi_pblock_new();
+    slapi_search_internal_set_pb_ext(pb, sdn, LDAP_SCOPE_BASE, ALIASFILTER, attrs, 0, NULL, NULL, plugin_id, 0);
+    slapi_pblock_set(pb, SLAPI_SEARCH_DEREF, (void *)&derefNever);
+
+    if (slapi_search_internal_pb(pb) == 0) {
+        slapi_pblock_get(pb, SLAPI_PLUGIN_INTOP_SEARCH_ENTRIES, &e);
+        if (e && e[0]) {
+            entry_sdn = slapi_sdn_new_dn_byval(slapi_entry_attr_get_ref(e[0], attrs[0]));
+        }
+    }
+    slapi_free_search_results_internal(pb);
+    slapi_pblock_destroy(pb);
+
+    return entry_sdn;
+}
+
+int
+alias_entry_init(Slapi_PBlock *pb)
+{
+    if (slapi_pblock_set(pb, SLAPI_PLUGIN_VERSION, SLAPI_PLUGIN_VERSION_03) != 0 ||
+        slapi_pblock_set(pb, SLAPI_PLUGIN_DESCRIPTION, (void *)&srchpdesc) != 0 ||
+        slapi_pblock_set(pb, SLAPI_PLUGIN_PRE_SEARCH_FN, (void *) alias_entry_srch) != 0  ||
+        slapi_pblock_get(pb, SLAPI_PLUGIN_IDENTITY, &plugin_id) != 0)
+    {
+        slapi_log_error(SLAPI_LOG_ERR, PLUGINNAME,
+                        "alias_entry_init: Error registering the plug-in.\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+int
+alias_entry_srch(Slapi_PBlock *pb)
+{
+    Slapi_DN *search_target = NULL;
+    Slapi_DN *dn1 = NULL;
+    Slapi_DN *dn2 = NULL;
+    char *str_filter = NULL;
+    int scope = 0;
+    int deref_enabled = 0;
+    size_t i = 0;
+
+    /* deref must be enabled */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_DEREF, &deref_enabled) != 0) {
+        return 0;
+    }
+    if (deref_enabled != LDAP_DEREF_FINDING && deref_enabled != LDAP_DEREF_ALWAYS) {
+        return 0;
+    }
+
+    /* Only base search is supported */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_SCOPE, &scope) != 0 || scope != LDAP_SCOPE_BASE) {
+        return 0;
+    }
+
+    /* Skip reserved operations */
+    if (slapi_op_reserved(pb)) {
+        return 0;
+    }
+
+    /* Base dn should be provided */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_TARGET_SDN, &search_target) != 0 || search_target == NULL) {
+        return 0;
+    }
+
+    /* Skip our own requests */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_STRFILTER, &str_filter) != 0 ||
+        strcmp(str_filter, ALIASFILTER) == 0)
+    {
+        return 0;
+    }
+
+    /* Follow the alias chain */
+    dn2 = search_target;
+    do {
+        dn1 = dn2;
+        dn2 = alias_get_next_dn(dn1);
+        if (i > 0 && dn2 != NULL) {
+            slapi_sdn_free(&dn1);
```

Comment:
I have a doubt here.

If the function alias_get_next_dn returns NULL, dn1 is not freed. And alias_get_next_dn can return NULL if (slapi_search_internal_pb(pb) != 0).

Reply:
So I wonder if we should change it to something like that:

```
    if (i > 0 && dn1 != search_target) {
        slapi_sdn_free(&dn1);
    }
```

So dn1 is always freed unless it is the original search_target.

Reply:
True, if the internal search failed (e.g. nonexistent DN in aliasedObjectName) then the parent in the chain would be returned instead.  So this can be hardened for a such a case.  

Do we log an warning?  
Do we return an error to the client?
Do we just return the parent?
Or return no result at all (err=0 nentries=0)?

Probably should just return an error to client: operations error?  Unwilling to perform?

Reply:
My vote is for the operations error. I think we should notify the user that something is wrong and point out that dereference is broken in his requested scope.

If the admin decided to use the plugin - probably there's a good reason for that and a warning, parent's return or just no results is not enough in my opinion.

@mreynolds389 @tbordaz @progier389 thoughts? :)

Reply:
The RFC does not provide any requirement for such a scenario.  I'm torn between not adjusting the search target at all, or returning an error (err=1 works for me).

Reply:
We can argue about it, but IMHO its depends of the failure cause.
  if it is NOSUCHOBJECT the current code behavior is not so bad: returning the last entry that we were able to dereference. Anyway by looking at its objectclass, the client application should easily determine that it is an alias and that something is wrong. (returning the original entry is possible too)
  If it is another error, I would return the internal search error code and a result text like: "'%s' while dereferencing alias %s" internal_search_result_text, dn1